src/opers.c

```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
-              cache[7*STATE(CacheIndex)+4] = id3;
-              cache[7*STATE(CacheIndex)+5] = id4;
-              cache[7*STATE(CacheIndex)+6] = id5;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
-            }
-#ifdef COUNT_OPERS
-          OperationMiss++;
-#endif
+
+            types[0] = FLAGS_FILT(arg1);
         }
-      if ( !method )  {
-        ErrorQuit( "no method returned", 0L, 0L );
-      }
-      
-      /* call this method                                                    */
-      res = CALL_5ARGS( method, arg1, arg2, arg3, arg4, arg5 );
+        else
+            types[0] = TYPE_OBJ_FEO(arg1);
+    case 0:
+        break;
+    default:
+        GAP_ASSERT(0);
     }
-    while (res == TRY_NEXT_METHOD );
-
-    /* return the result                                                   */
-    return res;
-}
 
+    if (n > 0) {
+        if (constructor)
+            ids[0] = types[0];
+        else
+            ids[0] = ID_TYPE(types[0]);
+    }
 
+    for (UInt i = 1; i < n; i++)
+        ids[i] = ID_TYPE(types[i]);
 
-/****************************************************************************
-**
-**  DoOperation6Args( <oper>, <a1>, <a2>, <a3>, <a4>, <a5>, <a6> )
-*/
-Obj DoOperation6Args (
-    Obj                 oper,
-    Obj                 arg1,
-    Obj                 arg2,
-    Obj                 arg3,
-    Obj                 arg4,
-    Obj                 arg5,
-    Obj                 arg6 )
-{
-    Obj                 res;
-    Obj                 type1;
-    Obj                 id1;
-    Obj                 type2;
-    Obj                 id2;
-    Obj                 type3;
-    Obj                 id3;
-    Obj                 type4;
-    Obj                 id4;
-    Obj                 type5;
-    Obj                 id5;
-    Obj                 type6;
-    Obj                 id6;
-    Obj *               cache;
-    Obj                 method;
-    Obj                 margs;
-    Int                 i;
-    Obj                 prec;
-
-    /* get the types of the arguments                                      */
-    type1 = TYPE_OBJ_FEO( arg1 );
-    id1 = ID_TYPE( type1 );
-    type2 = TYPE_OBJ_FEO( arg2 );
-    id2 = ID_TYPE( type2 );
-    type3 = TYPE_OBJ_FEO( arg3 );
-    id3 = ID_TYPE( type3 );
-    type4 = TYPE_OBJ_FEO( arg4 );
-    id4 = ID_TYPE( type4 );
-    type5 = TYPE_OBJ_FEO( arg5 );
-    id5 = ID_TYPE( type5 );
-    type6 = TYPE_OBJ_FEO( arg6 );
-    id6 = ID_TYPE( type6 );
-    
-    /* try to find an applicable method in the cache                       */
+    /* outer loop deals with TryNextMethod */
     prec = INTOBJ_INT(-1);
-
     do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
-
-      /* recalculate cache each pass, in case of GC     */
-      cache = 1+ADDR_OBJ( CacheOper( oper, 6 ) );
+        prec = INTOBJ_INT(INT_INTOBJ(prec) + 1);
```

Comment:
Why not change `prec` to an `Int`? Most places using it seem to just run `INT_INTOBJ` on it anyway; only in `CacheMethod` is the `INTOBJ` value being used (and stored), but (a) only after checking its `Int` value (so it performs a conversion anyway), and (b) that is of course only called once, to record a hit -- so it would seem better to only convert once, instead of constantly converting back and forth, no? (Ah yes, `CallHandleMethodNotFound` also needs the OBJ value, but that, too, is only rarely called).

Reply:
It is stored in the cache, which is just a Plist, so what is stored there needs to be legal GAP object. It might be sensible to  convert it immediately on reading from there though.

Reply:
Of course inside the cache, it is stored as an intobj; but that doesn't mean we have to keep it in that representation while using it. :-)